text/0000-dynamic-tag-names.md

```diff
@@ -0,0 +1,96 @@
+- Start Date: 2018-10-14
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Dynamic tag names in glimmer templates.
+
+## Summary
+
+With the transition from inner-html semantics to outer-html semantics in components, we lost one feature: Being
+able dynamically define the tag name of components.
+
+I think it was an useful feature and we should find a way to bring it back.
+
+## Motivation
+
+Although not something we use every day, there is a need for some components to have a dynamic tag name.
+
+This is most often used for certain low-level _presentational_ components.
+
+Take for instance a component named `<Panel>` that is used to encapsulate some presentation concerns.
+The template of that component could be like this:
+
+```hbs
+<div class="pt-10 pb-10 ps-20 box-shadow" ...attributes>
+  {{yield}}
+</div>
+```
+
+For accessibility and semantic reasons, sometimes a `<div>` may not be the best kind of tag.
+We might want the panel to be a `<section>` element, or an `<aside>` when it's content is somewhat unrelated
+with the rest of the page. Or maybe a `<legend>` if it's and the end of a form.
+
+## Detailed design
+
+Unlike ids, classes or other attributes, the tag name of DOM element cannot be changed in runtime.
+This makes creating a component with a dynamic _bound_ tag name tricky, as every time the tag name changes
+all the content of the tag has to be removed and added again. That would mean that all the sub-components
+of that component would be torn down and instantiated again.
+
+To help the user understand that changing the tag of an element in runtime is an expensive operation
+we must choose a syntax that reflects that complexity.
+
+The proposal includes the creation of a new `element` helper that takes a tag name and generates a contextual component that,
+when invoked, renders the selected element.
+
+Example:
+
+```hbs
+{{#let (element "li") as |Tag|}}
+  <Tag ...>...</Tag>
+{{/let}}
+```
+Since the element name can be defined in runtime, in order to protect the user from XSS vulnerabilities,
+the passed tag name cannot contain anything but lowercase letters and dashes. Tag names containing
+other characters (p.e `xlink:`) will throw an error.
+
+A working proof of concept of this approach has been created in https://github.com/tildeio/ember-element-helper
+
+Shall this RFC me accepted, that helper would be ported to Ember.js itsef, perhaps making it more efficient
+on the process.
+
+## How we teach this
+
+If this RFC is accepted, Ember would get a new built-in helper to be added to `component`, `array`, `each` and in
+[Ember.Templates.helpers](https://emberjs.com/api/ember/release/classes/Ember.Templates.helpers).
```

Comment:
The documentation will need a at least somewhat meaningful and self-contained example to illustrate the point of the helper (and how to use it). Could you provide one? I assume you have some examples from your addons that could be distilled down.